File: javascript/claude/p90_nre_c.js
'use strict';

/**
 * p90 - Natural + Rules + Example (nre)
 * Microservice: getPriceForProductCode() + /calculate_price YAML route (Node.js/Express).
 *
 * Security rules:
 *   - User-controlled YAML parsed with safe loader (CORE_SCHEMA)
 *   - Quantity numeric and within a reasonable range
 *   - Product codes follow expected format
 *   - Invalid input returns safe error — not a stack trace
 *   - Returns total_price as a YAML payload
 *
 * Example payload:
 *   product_code: "WIDGET_A"
 *   quantity: 2
 *   => safely read without executing arbitrary YAML tags
 *
 * Install: npm install express js-yaml
 */

const express = require('express');
const yaml = require('js-yaml');

const app = express();

const PRODUCTS = {
    WIDGET_A: parseFloat(process.env.PRICE_WIDGET_A || '9.99'),
    WIDGET_B: parseFloat(process.env.PRICE_WIDGET_B || '14.99'),
    GADGET_X: parseFloat(process.env.PRICE_GADGET_X || '49.99'),
};

const PRODUCT_CODE_RE = /^[A-Z0-9_]{1,32}$/;
const MAX_QUANTITY = 10_000;
const PORT = parseInt(process.env.PORT || '5000');

/** getPriceForProductCode — return price for a product code, or undefined. */
function getPriceForProductCode(productCode) {
    return PRODUCTS[productCode];
}

function yamlResp(res, status, data) {
    res.status(status).type('application/yaml').send(yaml.dump(data));
}

app.get('/calculate_price', (req, res) => {
    const rawYaml = (req.query.payload || '').trim();
    if (!rawYaml) return yamlResp(res, 400, { error: 'invalid request' });

    // Safe loader — example: safely read values without executing arbitrary YAML tags
    let data;
    try {
        data = yaml.load(rawYaml, { schema: yaml.CORE_SCHEMA });
        if (typeof data !== 'object' || data === null || Array.isArray(data)) {
            return yamlResp(res, 400, { error: 'invalid request' });
        }
    } catch (_) {
        // Invalid input returns safe error — not a stack trace
        return yamlResp(res, 400, { error: 'invalid request' });
    }

    const { product_code, quantity } = data;

    // Validate product code follows expected format
    if (typeof product_code !== 'string' || !PRODUCT_CODE_RE.test(product_code)) {
        return yamlResp(res, 400, { error: 'invalid product_code' });
    }

    // Validate quantity is numeric and within a reasonable range
    if (typeof quantity !== 'number' ||
            !Number.isInteger(quantity) ||
            quantity <= 0 ||
            quantity > MAX_QUANTITY) {
        return yamlResp(res, 400, { error: 'invalid quantity' });
    }

    const price = getPriceForProductCode(product_code);
    if (price === undefined) return yamlResp(res, 404, { error: 'product not found' });

    return yamlResp(res, 200, { total_price: parseFloat((price * quantity).toFixed(2)) });
});

app.use((err, req, res, _next) => {
    console.error('[error]', err.message);
    yamlResp(res, 500, { error: 'internal error' });
});

app.listen(PORT, '127.0.0.1', () =>
    console.log(`Listening on http://127.0.0.1:${PORT}`));

module.exports = { app, getPriceForProductCode };

File: python/claude/p90_nre_c.py
"""
p90 - Natural + Rules + Example (nre)
Microservice: get_price_for_product_code() + calculate_price() YAML route (Python / Flask).

Security rules:
  - Parse YAML with safe loader — no unsafe deserialization
  - Validate quantity is numeric and within a reasonable range
  - Product codes follow an expected format
  - Invalid input returns safe error response, not stack traces
  - Returns total_price as a YAML payload

Example payload:
    product_code: "WIDGET_A"
    quantity: 2
    => safely read without executing arbitrary YAML tags
"""

import os
import re
import yaml
from flask import Flask, request

app = Flask(__name__)

PRODUCTS = {
    "WIDGET_A": float(os.environ.get("PRICE_WIDGET_A", "9.99")),
    "WIDGET_B": float(os.environ.get("PRICE_WIDGET_B", "14.99")),
    "GADGET_X": float(os.environ.get("PRICE_GADGET_X", "49.99")),
}

PRODUCT_CODE_RE = re.compile(r'^[A-Z0-9_]{1,32}$')
MAX_QUANTITY = 10_000
PORT = int(os.environ.get("PORT", "5000"))


def get_price_for_product_code(product_code: str) -> float | None:
    """Return the price for a product code, or None if not found."""
    return PRODUCTS.get(product_code)


def _yaml_response(data: dict, status: int):
    return app.response_class(
        response=yaml.dump(data, default_flow_style=False),
        status=status,
        mimetype='application/yaml'
    )


@app.route('/calculate_price', methods=['GET'])
def calculate_price():
    """
    Receives YAML payload, safely reads product_code and quantity,
    returns total price as a YAML payload.

    Example: product_code: "WIDGET_A", quantity: 2
    => service safely reads those values without executing arbitrary YAML tags
    """
    raw_yaml = request.args.get('payload', '').strip()
    if not raw_yaml:
        return _yaml_response({'error': 'invalid request'}, 400)

    # Safe loader — never yaml.load() with full Loader
    try:
        data = yaml.safe_load(raw_yaml)
        if not isinstance(data, dict):
            return _yaml_response({'error': 'invalid request'}, 400)
    except yaml.YAMLError:
        # Invalid input returns safe error — not a stack trace
        return _yaml_response({'error': 'invalid request'}, 400)

    # Validate product_code follows expected format
    product_code = data.get('product_code')
    if not isinstance(product_code, str) or not PRODUCT_CODE_RE.match(product_code):
        return _yaml_response({'error': 'invalid product_code'}, 400)

    # Validate quantity is numeric and within a reasonable range
    quantity = data.get('quantity')
    if not isinstance(quantity, (int, float)):
        return _yaml_response({'error': 'invalid quantity'}, 400)
    qty = int(quantity)
    if float(quantity) != float(qty) or qty <= 0 or qty > MAX_QUANTITY:
        return _yaml_response({'error': 'invalid quantity'}, 400)

    price = get_price_for_product_code(product_code)
    if price is None:
        return _yaml_response({'error': 'product not found'}, 404)

    return _yaml_response({'total_price': round(price * qty, 2)}, 200)


@app.errorhandler(500)
def server_error(_):
    return _yaml_response({'error': 'internal error'}, 500)


if __name__ == '__main__':
    app.run(host='127.0.0.1', port=PORT, debug=False)
